lib/open_project/scm/adapters/subversion.rb

```diff
@@ -0,0 +1,289 @@
+        end
+
+        def xml_capture(cmd, opts = {})
+          output = capture_svn(cmd, opts)
+          doc = Nokogiri::XML(output)
+
+          # Yield helper methods instead of doc
+          yield doc
+        end
+
+        def extract_entry(entry, path)
+          revision = extract_revision(entry.at_xpath('commit'))
+          kind, size, name = parse_entry(entry)
+
+          # Skip directory if there is no commit date (usually that
+          # means that we don't have read access to it)
+          return if kind == 'dir' && revision.time.nil?
+
+          Entry.new(
+            name: URI.unescape(name),
+            path: ((path.empty? ? '' : "#{path}/") + name),
+            kind: kind,
+            size: size.nil? ? nil : size.to_i,
+            lastrev: revision
+          )
+        end
+
+        def parse_entry(entry)
+          kind = entry['kind']
+          size = entry.xpath('size').text
+          name = entry.xpath('name').text
+
+          [kind, size, name]
+        end
+
+        def build_path(entry)
+          {
+            action: entry['action'],
+            path: entry.text,
+            from_path: entry['copyfrom-path'],
+            from_revision: entry['copyfrom-rev']
+          }
+        end
+
+        def extract_revision(commit_node)
+          # We may be unauthorized to read the commit date
+          date =
+            begin
+              Time.parse(commit_node.xpath('date').text).localtime
+            rescue ArgumentError
+              nil
+            end
+
+          Revision.new(
+            identifier: commit_node['revision'],
+            time: date,
+            message: commit_node.xpath('msg').text,
+            author: commit_node.xpath('author').text
+          )
+        end
+
+        def fetch_revision_entries(identifier_from, identifier_to, options, path, &block)
+          path ||= ''
+          identifier_from = numeric_identifier(identifier_from, 'HEAD')
+          identifier_to = numeric_identifier(identifier_to, 1)
+          cmd = ['log', '--xml', '-r', "#{identifier_from}:#{identifier_to}"]
+          cmd << '--verbose' if options[:with_paths]
+          cmd << '--limit' << options[:limit].to_s if options[:limit]
+          cmd << target(path)
+          xml_capture(cmd, force_encoding: true) do |doc|
+            doc.xpath('/log/logentry').each &block
+          end
+        end
+
+        def target(path = '')
+          base = path.match(/\A\//) ? root_url : url
+          uri = "#{base}/#{path}"
+          URI.escape(URI.escape(uri), '[]')
+          # shell_quote(uri.gsub(/[?<>\*]/, ''))
+        end
+
+        ##
+        # Builds the full git arguments from the parameters
+        # and return the executed stdout as a string
+        def capture_svn(args, opt = {})
+          cmd = build_svn_cmd(args)
+          output = capture_out(cmd)
+
+          if opt[:force_encoding] && output.respond_to?(:force_encoding)
+            output.force_encoding('UTF-8')
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.